GitLab CI testing expansion (#1644)
LLNL/RAJA: docs/sphinx/dev_guide/ci_tasks.rst

   describe the change in the ``RAJA/.gitlab/jobs/<MACHINE>.yml`` file where
   the job is overridden.

Building the Intel clang + SYCL HIP compiler for use in CI
------------------------------------------------------

The SYCL CI tests on corona rely on a custom Intel Clang SYCL compiler that we 

> build ourselves. This compiler lives in the /usr/workspace/raja-dev/ folder so 

Reviewer: ```suggestion
build ourselves. This compiler lives in the ``/usr/workspace/raja-dev`` folder so 
```

that it can be accessed by the gitlab CI system. Since the intel compiler does
not do releases in the typical sense (they simply update their repo *every night*), 
it may become necessary to periodically build a new version of the compiler to 
ensure that we are using the most up-to-date version available. The steps for 

> building/installing/running are shown here.

Reviewer: ```suggestion
building, installing, and running are shown here.
```

Building the Compiler
^^^^^^^^^^^^^^^^^^^^^

> .. important:: Because intel update's their compiler repo daily, there is a nonzero possibility that the had of the sycl branch will build. 

Reviewer 2: ```suggestion
.. important:: Because intel update's their compiler repo daily, there is a nonone possibility that the head of the sycl branch will build. 
```

Author: Right, I guess I had the logic backwards. I'm going to change it to nonzero chance the head of develop will _fail to_ build

> .. important:: Because intel update's their compiler repo daily, there is a nonzero possibility that the had of the sycl branch will build. 

Reviewer: ```suggestion
.. important:: Because intel updates their compiler repo daily, there is a nonzero possibility that the had of the sycl branch will build. 
```

>   In the event that it does not build, try checking out a different commit. On the intel/llvm github page, one can see which of their 

Reviewer: ```suggestion
  In the event that it does not build, try checking out a different commit. On the intel/llvm GitHub page, one can see which of their 
```

  commits builds by checking the status badge next to each commit. Look for a commit that passes. 

#. 1. Load the version of GCC that you want to use. in this case, we are using LC's gcc/10.3.1-magic installation. 

  .. code-block:: bash

    module load gcc/10.3.1-magic

#. 2. Clone the "sycl" branch of intel's llvm compiler fork 

  .. code-block:: bash

    git clone https://github.com/intel/llvm -b sycl

#. 3. cd into that folder 

  .. code-block:: bash
    
    cd llvm

  #. 3a. In the event that the head of the sycl branch does not build, run `git checkout <git sha>` to checkout a version that does build.

> #. 4. build the compiler. Note that in this example, we are using rocm5.7.1, but one can change the version they wish to use simply by changing the paths in the configure step.

Reviewer: Does  the rocm version module need to be loaded? I think it does, but I may be wrong.

Author: Oh yeah, you're right, it does

  #. a. Configure

    .. code-block:: bash 

      srun -n1 /usr/bin/python3 buildbot/configure.py --hip -o buildrocm5.7.1 \
      --cmake-gen "Unix Makefiles" \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_ROCM_DIR=/opt/rocm-5.7.1 \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_ROCM_INCLUDE_DIR=/opt/rocm-5.7.1/include \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_ROCM_LIB_DIR=/opt/rocm-5.7.1/lib \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_INCLUDE_DIR=/opt/rocm-5.7.1/include \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_HSA_INCLUDE_DIR=/opt/rocm-5.7.1/hsa/include/hsa \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_LIB_DIR=/opt/rocm-5.7.1/lib \
      --cmake-opt=-DUR_HIP_ROCM_DIR=/opt/rocm-5.7.1 \
      --cmake-opt=-DUR_HIP_INCLUDE_DIR=/opt/rocm-5.7.1/include \
      --cmake-opt=-DUR_HIP_HSA_INCLUDE_DIR=/opt/rocm-5.7.1/hsa/include/hsa \
      --cmake-opt=-DUR_HIP_LIB_DIR=/opt/rocm-5.7.1/lib
  
  #. b. Build

    .. code-block:: bash 

      srun -n1 /usr/bin/python3 buildbot/compile.py -o buildrocm5.7.1

#. 5. Test the compiler

  Follow the steps in the next section "Using the compiler" to test this installation

#. 6. Install

>   #. a The build step will install the compiler to the folder buildrocm<version>/install. Simply copy this folder to the /usr/workspace/raja-dev/ 

Reviewer: ```suggestion
  #. a The build step will install the compiler to the folder ``buildrocm<version>/install``. Simply copy this folder to the ``/usr/workspace/raja-dev/``
```

>   directory using the naming scheme `clang_sycl_<git sha>_hip_gcc<version>_rocm<version>`

Reviewer: ```suggestion
  directory using the naming scheme ``clang_sycl_<git sha>_hip_gcc<version>_rocm<version>``
```

>   #. b set the permissions of the folder, and everything in it to 750

Reviewer 2: do the groups need to change as well?

    .. code-block:: bash 

      chmod 750 /usr/workspace/raja-dev/<foldername>/ -R  

Using the compiler
^^^^^^^^^^^^^^^^^^

> #. 1. Load the version of rocm that you used when building the compiler

Reviewer: Same comment as above about numbered lists.

  .. code-block:: bash
    module load rocm/5.7.1

> #. 2. Navigate to the root of your raja repo

Reviewer: ```suggestion
#. 2. Navigate to the root of your local checkout space of the RAJA repo.
```

   describe the change in the ``RAJA/.gitlab/jobs/<MACHINE>.yml`` file where
   the job is overridden.

Building the Intel clang + SYCL HIP compiler for use in CI
------------------------------------------------------

The SYCL CI tests on corona rely on a custom Intel Clang SYCL compiler that we 
build ourselves. This compiler lives in the ``/usr/workspace/raja-dev/`` folder so 
that it can be accessed by the gitlab CI system. Since the intel compiler does
not do releases in the typical sense (they simply update their repo *every night*), 
it may become necessary to periodically build a new version of the compiler to 
ensure that we are using the most up-to-date version available. The steps for 
building, installing, and running are shown here.

Building the Compiler
^^^^^^^^^^^^^^^^^^^^^

.. important:: Because intel updates their compiler repo daily, there is a nonzero possibility that the head of the sycl branch will fail to build. 
  In the event that it does not build, try checking out a different commit. On the intel/llvm GitHub page, one can see which of their 
  commits builds by checking the status badge next to each commit. Look for a commit that passes. 

> #. Load the version of GCC that you want to use. in this case, we are using LC's gcc/10.3.1-magic installation. 

Reviewer: I think all list items need to be indented from the main text so that the numbers generate properly. When they are at the same level as the main text, each item has number '1'. Please see https://raja.readthedocs.io/en/long58-ci_expansion/sphinx/dev_guide/ci_tasks.html

Author: Yeah, I saw that, I was in the process of changing it. It took a little bit for it to update after I pushed my changes.

Reviewer: I don't recall what the exact rules are. If you run into something, there is likely an example elsewhere in the docs where it is done correctly. Not the most helpful comment, but that's often how I fix problems in the docs.

#. Build the compiler. Note that in this example, we are using rocm5.7.1, but one can change the version they wish to use simply by changing the paths in the configure step.

  #. Configure

    .. code-block:: bash 

      srun -n1 /usr/bin/python3 buildbot/configure.py --hip -o buildrocm5.7.1 \
      --cmake-gen "Unix Makefiles" \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_ROCM_DIR=/opt/rocm-5.7.1 \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_ROCM_INCLUDE_DIR=/opt/rocm-5.7.1/include \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_ROCM_LIB_DIR=/opt/rocm-5.7.1/lib \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_INCLUDE_DIR=/opt/rocm-5.7.1/include \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_HSA_INCLUDE_DIR=/opt/rocm-5.7.1/hsa/include/hsa \
      --cmake-opt=-DSYCL_BUILD_PI_HIP_LIB_DIR=/opt/rocm-5.7.1/lib \
      --cmake-opt=-DUR_HIP_ROCM_DIR=/opt/rocm-5.7.1 \
      --cmake-opt=-DUR_HIP_INCLUDE_DIR=/opt/rocm-5.7.1/include \
      --cmake-opt=-DUR_HIP_HSA_INCLUDE_DIR=/opt/rocm-5.7.1/hsa/include/hsa \
      --cmake-opt=-DUR_HIP_LIB_DIR=/opt/rocm-5.7.1/lib
  
  #. Build

    .. code-block:: bash 

      srun -n1 /usr/bin/python3 buildbot/compile.py -o buildrocm5.7.1

#. Test the compiler

>   Follow the steps in the next section "Using the compiler" to test this installation

Reviewer: Please add a cross reference to the other section.

#. Install

  #. The build step will install the compiler to the folder ``buildrocm<version>/install``. Simply copy this folder to the ``/usr/workspace/raja-dev/`` directory using the naming scheme ``clang_sycl_<git sha>_hip_gcc<version>_rocm<version>``

>   #. Set the permissions of the folder, and everything in it to 750

Reviewer: ```suggestion
  #. Set the permissions of the folder, and everything in it to 750::
```

>     .. code-block:: bash 

Reviewer: ```suggestion
```

For simple code-like blocks, I think this syntax is simpler and easier to get formatting correct.  Same with the item below. It looks like there is something wrong with the list formatting/numbering as it is.

      chmod 750 /usr/workspace/raja-dev/<foldername>/ -R  

  #. Change the group of the folder and everything in it to raja-dev 

    .. code-block:: bash 

      chgrp raja-dev /usr/workspace/raja-dev/<foldername>/ -R  

Using the compiler
^^^^^^^^^^^^^^^^^^

> #. Load the version of rocm that you used when building the compiler

Reviewer: Indent the list item lines to get proper numbering.
